### PR TITLE
feat!: policy versioning scaffolding

### LINKED ISSUE #171: Feature request: Validate policy assignment name length
When creating a custom policy assignment, the name must be no more than 24 characters long. 
It could be helpful if this can be validated within the provider so that the error can be picked up during plan rather than at apply.

```
│ Error: Failed to create/update resource
│ 
│   with module.alz.module.alz_global.module.policy_assignment["global/Audit-OpenAI-ContentLogging"].azapi_resource.this,
│   on .terraform/modules/alz.alz_global/modules/azapi_helper/main.tf line 1, in resource "azapi_resource" "this":
│    1: resource "azapi_resource" "this" {
│ 
│ creating/updating Resource: (ResourceId
│ "/providers/Microsoft.Management/managementGroups/global/providers/Microsoft.Authorization/policyAssignments/Audit-OpenAI-ContentLogging-MB"
│ / Api Version "2024-04-01"): PUT
│ https://management.azure.com/providers/Microsoft.Management/managementGroups/global/providers/Microsoft.Authorization/policyAssignments/Audit-OpenAI-ContentLogging
│ --------------------------------------------------------------------------------
│ RESPONSE 400: 400 Bad Request
│ ERROR CODE: InvalidPolicyAssignmentName
│ --------------------------------------------------------------------------------
│ {
│   "error": {
│     "code": "InvalidPolicyAssignmentName",
│     "message": "The policy assignment name 'Audit-OpenAI-ContentLogging' is invalid. The policy assignment name length must not exceed '24' characters."
│   }
│ }
│ --------------------------------------------------------------------------------
│ 
```

### LINKED ISSUE #1715: Feature request: Validate policy assignment name length
When creating a custom policy assignment, the name must be no more than 24 characters long. 
It could be helpful if this can be validated within the provider so that the error can be picked up during plan rather than at apply.

```
│ Error: Failed to create/update resource
│ 
│   with module.alz.module.alz_global.module.policy_assignment["global/Audit-OpenAI-ContentLogging"].azapi_resource.this,
│   on .terraform/modules/alz.alz_global/modules/azapi_helper/main.tf line 1, in resource "azapi_resource" "this":
│    1: resource "azapi_resource" "this" {
│ 
│ creating/updating Resource: (ResourceId
│ "/providers/Microsoft.Management/managementGroups/global/providers/Microsoft.Authorization/policyAssignments/Audit-OpenAI-ContentLogging-MB"
│ / Api Version "2024-04-01"): PUT
│ https://management.azure.com/providers/Microsoft.Management/managementGroups/global/providers/Microsoft.Authorization/policyAssignments/Audit-OpenAI-ContentLogging
│ --------------------------------------------------------------------------------
│ RESPONSE 400: 400 Bad Request
│ ERROR CODE: InvalidPolicyAssignmentName
│ --------------------------------------------------------------------------------
│ {
│   "error": {
│     "code": "InvalidPolicyAssignmentName",
│     "message": "The policy assignment name 'Audit-OpenAI-ContentLogging' is invalid. The policy assignment name length must not exceed '24' characters."
│   }
│ }
│ --------------------------------------------------------------------------------
│ 
```

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
This PR scaffolds support for policy versioning by bumping the armpolicy package to v0.10.0 and adding new assets for handling versioned policy definitions.

This pull request introduces several significant changes, including the addition of a generic versioned policy collection, enhancements to policy assignment validation, and updates to configuration and code style. For the most part, these changes to not impact the production code.

### Breaking Changes

* policy assignment name, displayname and description length now checked here rather than waiting for a failure from Azure API. This isn't breaking as such, just moving the error left.

### Dependencies

* Updating armpolicy to v0.10.0
* Adding the Masterminds/semver/v3 dependency

### Generic Versioned Policy Collection

* Support for policy definitions and policy set definitions, with version
* Added a new `VersionedPolicyCollection` generic type in `assets/genericVersionCollection.go` to manage versioned policies with methods for adding and retrieving versions. This includes validation for version constraints and ensuring consistency in policy names.
* Introduced comprehensive unit tests in `assets/genericVersionCollection_test.go` to validate the functionality of `VersionedPolicyCollection`, covering scenarios such as adding versioned and versionless policies, handling invalid versions, and retrieving policies by constraints.

### Policy Assignment Enhancements

* Added validation logic in `assets/policyAssignment.go` to ensure policy assignments meet specific criteria, such as non-nil fields and length constraints for names, display names, and descriptions. Introduced constants for maximum lengths. [[1]](diffhunk://#diff-b15b38e6e358b553d4149961b26be7d70245768fbac5ae4e63d0812e0405eec4R9-L32) [[2]](diffhunk://#diff-b15b38e6e358b553d4149961b26be7d70245768fbac5ae4e63d0812e0405eec4R68-R140)
* Implemented a new constructor `NewPolicyAssignmentValidate` to create and validate `PolicyAssignment` objects, ensuring assignments are valid upon creation.
* Added unit tests in `assets/policyAssignment_test.go` to verify the validation logic for `PolicyAssignment`, ensuring comprehensive coverage for edge cases like missing or invalid fields.
* fixes Azure/Azure-Landing-Zones#1715 

### Refactoring and Code Style

* Refactored processor to use assets package for policy assignments & simplified logic in `alzlib.go` by directly assigning values to `policyAssignments` without creating new objects, reducing unnecessary object creation.
* Updated `.editorconfig` to enforce consistent code style by adding rules to trim trailing whitespace and insert final newlines.

Future PRs will build on this work to modify the core package to support these new assets.

### CODE REVIEW

File: assets/policyDefinitionVersions.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	// Add versioned definition if there is no versionless definition
+	if pdvs.versionlessDefinition != nil {
+		return fmt.Errorf("PolicyDefinitionVersions.Add: cannot add versioned definition for policy %s when versionless definition already exists", displayName)
+	}
+
+	if _, ok := pdvs.versions[*sv]; ok {
+		return fmt.Errorf("PolicyDefinitionVersions.Add: version %s for policy %s already exists", *verStr, displayName)
+	}
+
+	for v := range maps.Values(pdvs.versions) {
```
[the remark this answers is not in the document]
Comment (matt-FFFFFF) [reply assets/policyDefinitionVersions.go:68]: Thanks but no thanks

**Review by jaredfholgate — APPROVED** [review]
LGTM
